keps/104-StartupPolicy/README.md

```diff
@@ -0,0 +1,420 @@
+              - -xc
+              - |
+                sleep 100
+  - name: driver
+    replicas: 2
+    template:
+      spec:
+        backoffLimit: 0 
+        completions: 2
+        parallelism: 2
+        template:
+          spec:
+            containers:
+            - name: driver
+              image: bash:latest
+              command:
+              - bash
+              - -xc
+              - |
+                sleep 10
+  - name: worker
+    replicas: 2
+    template:
+      spec:
+        backoffLimit: 0 
+        completions: 2
+        parallelism: 2
+        template:
+          spec:
+            containers:
+            - name: worker
+              image: bash:latest
+              command:
+              - bash
+              - -xc
+              - |
+                sleep 10
+```
+
+### Notes/Constraints/Caveats (Optional)
+
+In the JobSet API, ReplicatedJobs are the individual Job elements that are created.  These are Replicated so you can have the same Job replicated.  This adds some complexity in the concept of Ready as we must consider the overall ReplicatedJob Ready rather than using the individual Jobs.
+
+For example, a Job with a `Parallelism`: 4 and `Ready` of 3. Means that 1 pod of a Job is not ready.  
+The Job API has no concept of an overall readiness condition based on a percentage of pods that are ready.  
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+### API Proposal
+
+Keeping the non specified jobs halted we propose to use suspend for these jobs.
+
+So we will add a new field to the ReplicatedJobStatus.
+
+```golang
+// ReplicatedJobStatus defines the observed ReplicatedJobs Readiness.
+type ReplicatedJobStatus struct {
+  ...
+	Suspended int32  `json:"suspended"`
+}
+```
+
+```golang
+// Status of ReplicatedJob for StartupPolicy
+type Status string
+
+const (
+ // Ready means that the job is ready
+ Ready Status = "Ready"
+
+ +// Suspended means that the job is suspended
+ +Suspended Status = "Suspended"
+)
+
+type StartupPolicy struct {
+  // If this field is set to true, we will start all replicatedJobs in order.
+	// Started means that all the Jobs of a ReplicatedJob are ready.
+	// +optional
+	ReplicatedJobsStartInOrder *bool `json:"replicatedJobsStartInOrder,omitempty"`
+}
+```
+
+### Implementation
+
+If StartupPolicy is set and suspend field is off, we will start all jobs in a suspended state.
+
+We will not set a condition for JobSet as suspended so we can distinguish between a suspended JobSet and a JobSet with startup policy.
+To track which ReplicatedJobs are suspended, we will add a `Suspended` field to the ReplicatedJobStatus.  
```

Comment:
This will be confusing, can we use a different name? `WaitingToStart` or something along those lines. Also, when are you going to set this value? it needs to exist from creation time, so I think we need to do it in the defaulting webhook, but I am not sure if status can be set there.

Reply:
I was thinking it would be nice to have a status field for suspended jobs anyway. I halt the execution of the jobs using suspend so it made sense to title this suspended. I can mark it as `WaitingToStart`. The main issue is what do you want for jobs that are all suspended due to the `suspend` field being set?

I have a PR up for this https://github.com/kubernetes-sigs/jobset/pull/250. This mainly adds a new field for JobSet to allow for one to know how many ReplicatedJobs are suspended. And I was able to use this for the implementation PR.

Reply:
I have an implementation of this https://github.com/kubernetes-sigs/jobset/pull/246 if you allow for setting suspend when creating the jobs

Reply:
> I was thinking it would be nice to have a status field for suspended jobs anyway.

ok, we can add that, but suspend and startup sequence are two different concepts and I think it is a mistake to conflate them.

It seems that what we need here is a status to track which replicatedJob we started last to kickoff the next one. This should be separate from the concept of "suspend". We could use a single status field indicating the name of the replicatedJob we processed last.

Reply:
Do you think that this could just be a label? I don't see why it should be a status field if its an internal implementation detail.

Reply:
An annotation would be appropriate.

Reply:
I really don't know how we could use annotations to achieve this. The reason I went with Suspend is that there is really no other way in kube to halt the execution of a job once it is created. We have to define the jobs and create them at some point. So I think that starting these jobs in a suspended state is probably the only way to do this with what we have currently.

There is the job readiness gates that @alculquicondor is proposing that could be used here if we don't want to use suspend. But the problem is the same.

We want to create all the jobs in a JobSet and then start them in order. We can't really run the jobs in an order due to the async nature of controllers. It also would make applying failure-policy or success-policy difficult..

Reply:
We have three cases where we startupPolicy applies:

1) when the jobset is first created
2) when the jobset is restarted after a failure (with restart or recreate all failure policy)
3) when the jobset is resumed after suspension


In cases 1 and 2, the operator controls the order of creating the jobs, and so it creates the replicatedJob only when its turn comes in the startup sequence. In case 3, the operator controls the order of unsuspending the jobs, and so it unsuspends the replicatedJob only when its turn comes in the startup sequence.

We use the annotation to track which replicatedJob is next to create or unsuspend.

Reply:
Thanks for clearing that up. I updated the design to call this out explictly.